File: csharp/utils/Geometry/Position3D.cs
using System;
using System.Numerics;

namespace AOC_Utils.Geometry;

public struct Position3D<T> : IEquatable<Position3D<T>>
    where T : INumber<T>
{
    public T X { get; set; }
    public T Y { get; set; }
    public T Z { get; set; }

    public Position3D(T x, T y, T z)
    {
        X = x;
        Y = y;
	Z = z;
    }

    public Position3D<T> Add(Position3D<T> other)
    {
        return new Position3D<T>(X + other.X, Y + other.Y, Z + other.Z);
    }

    public override bool Equals(object? obj) => obj is Position3D<T> other && Equals(other);
    public bool Equals(Position3D<T> other) => X == other.X && Y == other.Y && Z == other.Z;

    public override int GetHashCode() => HashCode.Combine(X, Y, Z);
    public override string ToString() => $"Position3D({X}, {Y}, {Z})";
}

File: fsharp/utils/Geometry/Position3D.fs
namespace AOC_Utils.Geometry

open System
open System.Numerics

[<Struct>]
type Position3D<'T when 'T :> INumber<'T> and 'T : equality> =
    val X: 'T
    val Y: 'T
    val Z: 'T

    new(x: 'T, y: 'T, z: 'T) = { X = x; Y = y; Z = z }

    member this.Add(other: Position3D<'T>) =
        Position3D<'T>(this.X + other.X, this.Y + other.Y, this.Z + other.Z)

    override this.ToString() =
        $"Position3D({this.X}, {this.Y}, {this.Z})"
